syntax = "proto3";
package connection_request;

message NodeAddress {
    string host = 1;
    uint32 port = 2;
}

enum ReadFrom {
    Primary = 0;
    PreferReplica = 1;
    LowestLatency = 2;
    AZAffinity = 3;
}

enum TlsMode {
    NoTls = 0;
    SecureTls = 1;
    InsecureTls = 2;
}

message AuthenticationInfo {
    string password = 1;
    string username = 2;
}

enum ProtocolVersion {
    RESP3 = 0;
    RESP2 = 1; 
}

message PeriodicChecksManualInterval {
    uint32 duration_in_sec = 1;
}

message PeriodicChecksDisabled {
}

enum PubSubChannelType {
    Exact = 0;
    Pattern = 1;
    Sharded = 2;
}

message PubSubChannelsOrPatterns
{
    repeated bytes channels_or_patterns = 1;
}

message PubSubSubscriptions
{
    map<uint32, PubSubChannelsOrPatterns> channels_or_patterns_by_type = 1;
}

// IMPORTANT - if you add fields here, you probably need to add them also in client/mod.rs:`sanitized_request_string`.
message ConnectionRequest {
    repeated NodeAddress addresses = 1;
    TlsMode tls_mode = 2;
    bool cluster_mode_enabled = 3;
    uint32 request_timeout = 4;
    ReadFrom read_from = 5;
    ConnectionRetryStrategy connection_retry_strategy = 6;
    AuthenticationInfo authentication_info = 7;
    uint32 database_id = 8;
    ProtocolVersion protocol = 9;
    string client_name = 10;
    oneof periodic_checks {
        PeriodicChecksManualInterval periodic_checks_manual_interval = 11;
        PeriodicChecksDisabled periodic_checks_disabled = 12;
    }
    PubSubSubscriptions pubsub_subscriptions = 13;
}

message ConnectionRetryStrategy {
    uint32 number_of_retries = 1;
    uint32 factor = 2;
    uint32 exponent_base = 3;
}
